[Frontend] As a user, I can register

carryall opened the pull request:
## What happened

Resolved #23 
- use icon sprite with SVG icons
- add additional template functions
- add style for the default template, registration form, and alert component

## Insight

Describe in detail how to test the changes. Referenced documentation is welcome as well.

## Proof Of Work

_Signup screen_

| Small Screens | Large screens |
|-|-|
| <img width="332" alt="Screen Shot 2564-01-25 at 12 46 02" src="https://user-images.githubusercontent.com/1772999/105666383-b0c65900-5f0b-11eb-93e5-3aec74069208.png"> | <img width="870" alt="Screen Shot 2564-01-25 at 12 45 50" src="https://user-images.githubusercontent.com/1772999/105666358-a4da9700-5f0b-11eb-8118-b2f8b4fb767b.png"> |

_Alerts_
| Small Screens | Large screens |
|-|-|
| <img width="336" alt="Screen Shot 2564-01-25 at 12 47 26" src="https://user-images.githubusercontent.com/1772999/105666400-b6bc3a00-5f0b-11eb-8579-7c685ccb0c90.png"> | ![Screen Shot 2564-01-22 at 11 19 28](https://user-images.githubusercontent.com/1772999/105666132-2c73d600-5f0b-11eb-87a4-04ddd6dee8a1.png) |

_Submitting the registration form_

| Success | Error |
|-|-|
| ![go_user_register_form_success](https://user-images.githubusercontent.com/1772999/105666467-e1a68e00-5f0b-11eb-964f-896e739a2259.gif) | ![go_user_register_form_fail](https://user-images.githubusercontent.com/1772999/105666474-e79c6f00-5f0b-11eb-97da-502c29858ab1.gif) |

carryall commented on initializers/assets.go:
```diff
@@ -0,0 +1,10 @@
+
+// SetupStaticPaths set static paths for assets
+func SetupStaticPaths() {
+	web.SetStaticPath("/css", "static/css")
```
set these static parts to be able to use `css/*`, `js/*`, `svg/*` in template

gutakk commented on assets/stylesheets/index.scss:
```diff
@@ -1,6 +1,13 @@
 // Base
-@import 'base/header';
+@import 'base/form';
```
Do we need to have prefix for imported file?
```suggestion
@import 'base/_form';
```
And also the file name, should be `_form.scss`?

gutakk commented on views/users/new.tpl:
```diff
@@ -1,8 +1,35 @@
-  {{ .flash }}
+{{ range $flashType, $flashMessage := .flash }}
+  {{ if $flashMessage }}
+    <div class="alert alert--{{$flashType}}">
```
Can we move alert to something like `partial`? I think alert will be used more than one places.

gutakk commented on initializers/view.go:
```diff
@@ -0,0 +1,51 @@
+		log.Fatal("Failed to add template function", err.Error())
+	}
+
+	err = web.AddFuncMap("sentencecase", helpers.ToSentenceCase)
```
Should we change the `sentencecase` to `sentence_case` to make it consistent with `render_file/icon`?

olivierobert commented on Dockerfile:
```diff
@@ -14,10 +14,15 @@ RUN npm run build-scss
 
+# Generate SVG Sprite
+RUN npm install svg-sprite -g
+RUN svg-sprite -cD static -cscss assets/images/icons/*.svg
```
Why not using the makefile command `assets/icon-sprite` or the NPM script `npm run generate-svg-sprite`?

carryall replied:
@olivierobert updated on [412cc5d](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/412cc5d70f3d9bf049595eb32b8c7bef12acd357) to use make command

olivierobert commented on assets/stylesheets/components/_app-container.scss:
```diff
@@ -0,0 +1,3 @@
+.app-container {
```
These styles should go in the layout as they affect how the page rendered.

carryall replied:
@olivierobert ✅  moved on [ababbb6](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/ababbb635ed9dacd53a66508770b72b570b7dec5)

olivierobert replied:
That's not what I meant 🙃 There should be a layout file e.g. `_application.scss`, `_authentication.scss`. Then in the template file, on teh root HTML element add a class name:

```
<htm class="layout-application">
```

And use in the screen stylesheet:

```
.layout-application {
  .app-header {}
  .app-container {}
  ...
}
```

Recheck how this is done on any past project if you are unsure

olivierobert commented on assets/stylesheets/components/_app-content.scss:
```diff
@@ -0,0 +1,3 @@
+.app-content {
```
Ditto here: these styles should go in the layout as they affect how the page rendered. So `.app-content` can have a specific style for each layout.

olivierobert commented on assets/stylesheets/components/_app-header.scss:
```diff
@@ -0,0 +1,7 @@
+.app-header {
+  @apply max-w-md w-full;
```
Ditto here: these styles should go in the layout. But the title part can be in the component.

olivierobert commented on assets/stylesheets/components/_icon.scss:
```diff
@@ -0,0 +1,7 @@
+.icon-sprite {
+  display: none;
```
A more bulletproof way is to use the [`hidden`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/hidden) attribute in the layout:

```
<div class="icon-sprite" hidden="true">
```

carryall replied:
@olivierobert added [75a42c4](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/75a42c4bc17e92a14d8219b3ea5ecead06b7b6d6)

olivierobert commented on assets/stylesheets/pages/_user.scss:
```diff
@@ -0,0 +1,5 @@
+.user-controller {
```
Why using a directory named as `pages` instead of `screens` as per our conventions and as done by Tailwind as well 🤔

olivierobert commented on views/layouts/default.tpl:
```diff
@@ -5,16 +5,23 @@
-  <link type="text/css" rel="stylesheet" href="static/css/application.css" />
-  <link type="text/css" rel="stylesheet" href="static/css/tailwind.css" />
+  {{ assets_css "css/application.css" }}
+  {{ assets_css "css/tailwind.css" }}
```
1. Is `assets_css` from Beego?
2. Why is Tailwind imported separately and below the application styles?

carryall commented on views/layouts/default.tpl:
```diff
@@ -5,16 +5,22 @@
   <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
-  <link type="text/css" rel="stylesheet" href="static/css/application.css" />
-  <link type="text/css" rel="stylesheet" href="static/css/tailwind.css" />
+  {{ assets_css "/css/application.css" }}
```
@olivierobert Yes `assets_css` is from `Beego`, more template functions can be found [here](https://beego.me/docs/mvc/view/template.md), also include tailwind on [this commit](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/a1568436e39115ac2bbce49d090e1f843a6a244a) 🙏🏼

olivierobert commented on Dockerfile:
```diff
@@ -1,22 +1,20 @@
 FROM  node:14.15-alpine as assets-builder
 
 RUN apk --no-cache add ca-certificates
+RUN apk --no-cache add make
```
MInor thing I noticed but you might want to do to save of container layers:

```suggestion
RUN apk --no-cache add ca-certificates \
    apk --no-cache add make
```

Each instruction line like  `RUN`, `ADD` or `COPY` would create one distinct layer. On some platforms like [Heroku](https://devcenter.heroku.com/articles/container-registry-and-runtime#known-issues-and-limitations) there is a hard limit on the number of layers:

![image](https://user-images.githubusercontent.com/696529/107003929-1e05a400-67c0-11eb-805b-aa01b00df647.png)

carryall replied:
@olivierobert ✅  Fixed on [73d6280](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/73d62804c5db2804c93f12e64a033147a8154317)

olivierobert reviewed (changes requested):
This critical issue has yet to be addressed: https://github.com/carryall/go-google-scraper-challenge/pull/25#discussion_r569220512

![Screen Shot 2564-02-05 at 14 42 42](https://user-images.githubusercontent.com/696529/107004209-8a80a300-67c0-11eb-92e4-897ebb2c9d02.png)

carryall commented:
> This critical issue has yet to be addressed: [#25 (comment)](https://github.com/carryall/go-google-scraper-challenge/pull/25#discussion_r569220512)
> 
> ![Screen Shot 2564-02-05 at 14 42 42](https://user-images.githubusercontent.com/696529/107004209-8a80a300-67c0-11eb-92e4-897ebb2c9d02.png)

@olivierobert Updated on [4d0c3a8](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/4d0c3a8f9da82f80d715a3e4867df4005abf5ea4), I added a separate `authentication` layout for the signup page, moved the styles there but still keep the default layout for future usage. Also moved some common style for `app-container` to the component.

olivierobert commented on views/layouts/authentication.tpl:
```diff
@@ -0,0 +1,26 @@
+    {{ render_file "static/symbol/svg/sprite.symbol.svg" }}
+  </div>
+
+  <div class="app-container">
```
Avoid wrapping divs at all. This is an old technique that is not a best practice anymore since it cannot cater to all designs. Instead, prefer stacking each layout component and then add a container inside:

| To Avoid | Better |
|-|-|
| ![Screen Shot 2564-02-08 at 14 03 35](https://user-images.githubusercontent.com/696529/107186358-b09f8080-6a16-11eb-8d34-a248ae26dc6f.png) | ![Screen Shot 2564-02-08 at 14 02 05](https://user-images.githubusercontent.com/696529/107186377-bbf2ac00-6a16-11eb-939e-a0ab93224b49.png) |

carryall replied:
@olivierobert Fixed by removing `app-container` and moved paddings to `app-header` and `app-content` ` on [44b4d83](https://github.com/carryall/go-google-scraper-challenge/pull/25/commits/44b4d835a1cf95e6adf5078d251d42d628d72803)

olivierobert reviewed (commented):
Overall, apart from the layout issue, this PR seems good to go.

Moving forward, improve your PRs by organizing the screenshots into tables. I edited the description of this PR for you to see the improvement:

| Before | After |
|-|-|
| ![github com_carryall_go-google-scraper-challenge_pull_25(Laptop with MDPI screen)](https://user-images.githubusercontent.com/696529/107186544-0b38dc80-6a17-11eb-8fa4-09aaeb5542bd.png) | ![github com_carryall_go-google-scraper-challenge_pull_25(Laptop with MDPI screen) (1)](https://user-images.githubusercontent.com/696529/107186570-1855cb80-6a17-11eb-88f5-04903006d8cd.png) |

It's part of your efforts to improve PR descriptions.

carryall commented:
> Overall, apart from the layout issue, this PR seems good to go.
> 
> Moving forward, improve your PRs by organizing the screenshots into tables. I edited the description of this PR for you to see the improvement:
> 
> Before	After
> ![github com_carryall_go-google-scraper-challenge_pull_25(Laptop with MDPI screen)](https://user-images.githubusercontent.com/696529/107186544-0b38dc80-6a17-11eb-8fa4-09aaeb5542bd.png)	![github com_carryall_go-google-scraper-challenge_pull_25(Laptop with MDPI screen) (1)](https://user-images.githubusercontent.com/696529/107186570-1855cb80-6a17-11eb-88f5-04903006d8cd.png)
> It's part of your efforts to improve PR descriptions.

🤩  Thank you @olivierobert next time I'll put them into the tables.
